src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs

```diff
@@ -193,31 +206,53 @@ private bool ValidateWithdrawals(Block block, IReleaseSpec spec, out string? err
         return true;
     }
 
-    private bool ValidateBlobs(Block block, IReleaseSpec spec, out string? error)
+    private bool ValidateTransactionsDataGas(Block block, IReleaseSpec spec, out string? error)
     {
-        if (spec.IsEip4844Enabled && block.ExcessDataGas is null)
+        if (!spec.IsEip4844Enabled)
         {
-            error = "ExcessDataGas field is not set.";
-            if (_logger.IsWarn) _logger.Warn(error);
-            return false;
+            error = null;
+            return true;
         }
 
-        if (!spec.IsEip4844Enabled && block.ExcessDataGas is not null)
+        int blobsInBlock = 0;
+        UInt256 dataGasPrice = UInt256.Zero;
+        for (int txIndex = 0; txIndex < block.Transactions.Length; txIndex++)
         {
-            error = "ExcessDataGas field should not have value.";
-            if (_logger.IsWarn) _logger.Warn(error);
-            return false;
+            Transaction transaction = block.Transactions[txIndex];
+            if (!transaction.SupportsBlobs)
+            {
+                continue;
+            }
+            if (dataGasPrice == UInt256.Zero)
+            {
+                if (!DataGasCalculator.TryCalculateDataGasPricePerUnit(block.Header, out dataGasPrice))
+                {
+                    error = $"{nameof(dataGasPrice)} overflow.";
+                    if (_logger.IsWarn) _logger.Warn(error);
+                    return false;
+                }
+            }
+            if (transaction.MaxFeePerDataGas < dataGasPrice)
+            {
+                error = $"A transaction has unsufficient MaxFeePerDataGas {transaction.MaxFeePerDataGas} < {dataGasPrice}.";
+                if (_logger.IsWarn) _logger.Warn(error);
+                return false;
+            }
+
+            blobsInBlock += transaction.BlobVersionedHashes!.Length;
         }
 
-        int? blobsInBlock = 0;
-        for (int txIndex = block.Transactions.Length - 1; txIndex >= 0; txIndex--)
+        ulong dataGasUsed = DataGasCalculator.CalculateDataGas(blobsInBlock);
+        if (dataGasUsed > Eip4844Constants.MaxDataGasPerBlock)
         {
-            blobsInBlock += block.Transactions[txIndex].BlobVersionedHashes?.Length ?? 0;
+            error = $"A block cannot have more than {Eip4844Constants.MaxDataGasPerBlock} data gas.";
+            if (_logger.IsWarn) _logger.Warn(error);
+            return false;
         }
 
-        if (spec.IsEip4844Enabled && blobsInBlock > Eip4844Constants.MaxBlobsPerBlock)
+        if (dataGasUsed != block.Header.DataGasUsed)
         {
-            error = $"A block cannot contain more than {Eip4844Constants.MaxBlobsPerBlock} blobs.";
+            error = $"DataGasUsed does not match actual data gas used: {block.Header.DataGasUsed} != {dataGasUsed}.";
```

Comment:
minor detail: maybe "DataGasUsed `declared in BlockHeader` does not match actual data gas used: {block.Header.DataGasUsed} != {dataGasUsed}.